tools/clang/tools/libclang/dxcrewriteunused.cpp

```diff
@@ -502,7 +502,21 @@ ParsedSemanticDefineList hlsl::CollectSemanticDefinesParsedByCompiler(
   }
 
   if (!macros.empty()) {
-    MacroExpander expander(pp);
+    // Make a copy of the preprocessor. Since MacroExpander does lexing, the
+    // internal state of the preprocessor is modified as a result. Since we're
+    // calling this AFTER we have started parsing, messing with the internal
+    // state of the preprocessor will cause crashes later on, as more things
+    // are being parsed.
+    std::unique_ptr<PreprocessorOptions> Opts(
+        new PreprocessorOptions(pp.getPreprocessorOpts()));
+    clang::LangOptions langOptionsCopy = pp.getLangOpts();
+    std::unique_ptr<Preprocessor> ppCopy(new Preprocessor(
+        Opts.get(), compiler.getDiagnostics(), langOptionsCopy, pp.getSourceManager(),
+        pp.getHeaderSearchInfo(), pp.getModuleLoader(),
+        pp.getIdentifierTable().getExternalIdentifierLookup()));
+    Opts.release(); // Opts are owned by the preprocessor copy now.
```

Comment:
Should we use `Opts.release()` in the call above instead of `Opts.get()`?

Reply:
This is a pattern that I see elsewhere (like in `CompilerInstance::createPreprocessor`). The more I look at the called code though, the more I feel wrapping it in std::unique_ptr is not the right thing to do. I think the right way to do this is to just directly pass `new PreprocessorOptions`, or if the `new PreprocessorOptions` needs to be far away, use a `IntrusiveRefCntPtr` instead of a `std::unique_ptr` and calling release.

Reply:
The only reason I can think of to have the `release()` after the constructor would be if the constructor throws an exception before taking ownership of the pointer. I would consider that a bug in the called code. I don't think we have to change it, but it just looked a little weird to me.

Reply:
This whole section is gone. We're no longer making a copy of Preprocessor, because I realized it didn't actually work.